Admin notice modernisation

interlineae opened the pull request:
Fixes #2382

### Changes proposed in this Pull Request

* Different styling for different notice types added (error, warning, info, success)
* Setup for different notice classes in place
* Setup for icon retrieval in place
* Setup for Sensei's notice dismissal JS in place
* Setup for `get_dissmised_notices_`, `saved_dismissed_notices`, and `handle_notice_dimiss` is in place.

After discussing with @aaronfc about the scope of #2382 , we determined that setting up this PR, and the issue in question, would entail (1) firming up admin styling while (2) setting up persistent notice dismissal in a following issue. This was determined after realising that it would be a mark larger task than initially realised to have notice dismissal be persistent while also moving over all notice types to the new implementation.

To that end, this PR, like the issue it fixes:

* Adds support for different WPJM notice styling in admin
* Provides a foundation for notices to be migrated to the new notice design
* Makes notices dismissable, with persistence to be added in a latter, separate issue

### Testing instructions 

* Check out branch
* Ensure that you have an older version of a paid WPJM add-on active. If not, use a discount code and grab a license for an addon from https://wpjobmanger.com, then download an older copy of the addon from GH.
* You should then see a notice like so:

![notice](https://user-images.githubusercontent.com/2500940/232040947-aba405cb-b75c-4d2b-9f71-b81ffe468004.png)

interlineae committed:
Add dismiss button and styles

- Adds built in is-dismissable button
- Adds styling to vertically center button

interlineae committed:
Add dismiss button permanence

Add JS to ensure `is-dismissable` means that, and doesn't just mean "until you refresh"

interlineae committed:
Add notice border styling

Adds styling for the various levels of border notices

interlineae committed:
Shore up notice styling:

- Move `$notice-*` styles from `mixins` to `admin-notices.scss`
- Update styles to remove `sensei-` references
- Ensure styles load on **Job Listings --> Job Types**
- Bold notice text (except buttons) so that it matches the font weight of existing notices

interlineae committed:
Notice dismissal setup

- Copy functions from Sensei into WPJM to begin setup of persistent notice dismissal
- Ensure notice styling is styled on the main WP Admin dashboard page
- Prep for use of `$notice_id`
- Prep for inclusion of notice icon
- PHPCBF fixes

aaronfc commented on includes/admin/class-wp-job-manager-admin-notices.php:
```diff
@@ -317,16 +332,40 @@ private static function generate_notice_from_updates( $updates ) {
+
+		$notice_class = '';
+		if ( ! empty( $notice['style'] ) ) {
+			$notice_class = 'wpjm-notice-' . $notice['style'];
```
I think it would be nice to check if `style` is info, error, warning, success and fall back to info.

interlineae replied:
I've committed a change for this at https://github.com/Automattic/WP-Job-Manager/pull/2401/commits/4c798a9f95176ca5b17fa66c93fa1d2ab6157a44.

interlineae replied:
@aaronfc I've made another change to hopefully improve handling of this, so that we don't risk nonsense CSS classes. https://github.com/Automattic/WP-Job-Manager/pull/2401/commits/03898a95f2a118ce6de45554cbc7950238abc153

aaronfc commented on includes/admin/class-wp-job-manager-admin-notices.php:
```diff
@@ -317,16 +332,40 @@ private static function generate_notice_from_updates( $updates ) {
+			$notice_class = 'wpjm-notice-' . $notice['style'];
+		}
+
+		$is_dismissible       = $notice['dismissible'];
```
```suggestion
		$is_dismissible       = $notice['dismissible'] ?? true;
```

Not sure if true or false by default. Since false is more restrictive I would say true... but I am open for discussion.

interlineae replied:
At the moment, `$is_dismissable` is `null`. If the endpoint sets that to an empty string it'll be `false`/`0`, but with the understanding that we want these notices to _always_ be dismissible, then I agree with defaulting to `true`. That way, a non-dismissible notice would also have to be intentional.

aaronfc commented on includes/admin/class-wp-job-manager-admin-notices.php:
```diff
@@ -317,16 +332,40 @@ private static function generate_notice_from_updates( $updates ) {
+			$notice_wrapper_extra = sprintf( ' data-dismiss-action="wpjm_dismiss_notice" data-dismiss-notice="%1$s" data-dismiss-nonce="%2$s"', esc_attr( $notice_id ), esc_attr( wp_create_nonce( self::DISMISS_NOTICE_NONCE_ACTION ) ) );
+		}
+
+		// TODO Remove hard-coded 'is-dismissable' CSS class once all notices are converted to use this class.
+		echo '<div class="notice wpjm-admin-notice wpjm-admin-notice--info is-dismissible' . esc_attr( $notice_class ) . '"' . esc_attr__( $notice_wrapper_extra ). '>';
```
I think that if we do the above change we can already remove the `is-dismissible` already.

interlineae replied:
That we can!

interlineae replied:
Committed https://github.com/Automattic/WP-Job-Manager/pull/2401/commits/adc630bf257f5104f22ae576597dc46e719f916b

aaronfc commented on includes/admin/class-wp-job-manager-admin-notices.php:
```diff
@@ -317,16 +332,40 @@ private static function generate_notice_from_updates( $updates ) {
 		echo '<div class="wpjm-admin-notice__top">';
-		// TODO Add icon.
+
+		// TODO Implement get_icon method.
+		if ( ! empty( $notice['icon'] ) ) {
+			// phpcs:disable WordPress.Security.EscapeOutput.OutputNotEscaped -- Dynamic parts escaped in the function.
+			echo get_icon( $notice['icon'], 'wpjm-notice__icon' );
```
Not sure I understand this. What is this `get_icon` method? Is a method that does not exist?
We need to implement this. I am happy if we initially only add a method like:

```
private function get_icon( $icon_name ) {
    switch ($icon_name) {
        case 'wpjm':
            return '';// path to the actual image
    }
    return ''; // default image (maybe the wpjm one?)
}
```

We can talk about this if you want over slack.

aaronfc replied:
If you think this will take a lot of time, we can just leave the todo (without the non-existing method) and create a separate ticket to implement this.

interlineae replied:
@aaronfc Let's chat about this, as thinking over it, I'm not sure that a `get_icon` method is required, but let's chat.

aaronfc replied:
Talked over slack and we will implement a private method for this and add the new icon asset.

interlineae replied:
@aaronfc I've committed for this in https://github.com/Automattic/WP-Job-Manager/pull/2401/commits/722258ccd6c7dc40366f9fcafaf1e200c343d0c0. I've kept the WPJM as a fallback icon for branding/recognisability. We can always return an empty string instead if we thought against that.

aaronfc commented on includes/admin/class-wp-job-manager-admin-notices.php:
```diff
@@ -317,16 +332,40 @@ private static function generate_notice_from_updates( $updates ) {
+
 		echo '<div class="wpjm-admin-notice__message">';
-		echo wp_kses( $notice['message'], self::ALLOWED_HTML );
+		echo '<strong>' . wp_kses( $notice['message'], self::ALLOWED_HTML ) . '</strong>';
```
maybe we want to add `font-weight: bold` to the `wpjm-admin-notice__message` better?

interlineae replied:
I had gone with `<strong>` for the semantic significance (i.e. this is an important notice for screen readers). Maybe that's a misplaced concern, and if so, I can switch this out to the `font-weight` property.

Or, is there another consideration between the two that I'm leaving out?

aaronfc replied:
🤔 I didn't thought about the semantic significance... makes sense to me now :)

aaronfc commented on includes/admin/class-wp-job-manager-admin-notices.php:
```diff
@@ -350,6 +389,37 @@ private static function render_notice( $notice ) {
+
+	}
+
+	/**
+	 * Get the dismissed notifications (either for the user or site-wide).
+	 *
+	 * @param bool $is_user_notification True if this is for a user notification (vs site-wide notification).
+	 *
+	 * @return array
+	 */
+	private function get_dismissed_notices( $is_user_notification ) {
+		// TODO Implement from Sensei.
+	}
+
+	/**
+	 * Save dismissed notices.
+	 *
+	 * @param array $dismissed_notices Array of dismissed notices.
+	 * @param bool  $is_user_notification True if we are setting user notifications (vs site-wide notifications).
+	 */
+	private function save_dismissed_notices( $dismissed_notices, $is_user_notification ) {
+		// TODO Implement from Sensei.
+	}
```
Can you remove these? Since we are not using them we don't really need these skeletons here.

interlineae replied:
Committed https://github.com/Automattic/WP-Job-Manager/pull/2401/commits/f82515acd805ecddeac7434e5f7a66de48f4c4ca

aaronfc commented on includes/admin/class-wp-job-manager-admin.php:
```diff
@@ -89,7 +89,7 @@ public function admin_enqueue_scripts() {
 
 		$screen = get_current_screen();
-		if ( in_array( $screen->id, apply_filters( 'job_manager_admin_screen_ids', [ 'edit-job_listing', 'plugins', 'job_listing', 'job_listing_page_job-manager-settings', 'job_listing_page_job-manager-addons' ] ), true ) ) {
+		if ( in_array( $screen->id, apply_filters( 'job_manager_admin_screen_ids', [ 'edit-job_listing', 'plugins', 'job_listing', 'job_listing_page_job-manager-settings', 'job_listing_page_job-manager-addons', 'edit-job_listing_type', 'dashboard' ] ), true ) ) {
```
Is this needed?

interlineae replied:
With the notices as they currently display, if this isn't added, then navigating to **Job Listings --> Job Types** or WP Admin home shows the notice as so:

![styles](https://user-images.githubusercontent.com/2500940/232081594-464a9c27-c410-467a-a3d5-6dd40289081c.png)

Did we want to hide the notices on these pages instead?

renatho replied:
About this, I'd say that it would be nice to add only the needed assets, otherwise, we add a lot of unneeded assets to the page.

But I see that it's the current standard here and it seems to be including all the assets used in the admin. And since it's in the admin, I think it's fine to keep that for now to avoid a big refactor.

aaronfc replied:
Thanks for the explanation @anaemnesis ! I am good with this approach too

aaronfc commented on webpack.config.js:
```diff
@@ -11,13 +11,14 @@ const files = {
 	'css/job-submission': 'css/job-submission.scss',
 	'css/setup': 'css/setup.scss',
-	'css/menu': 'css/menu.scss',
+	'css/menu': 'css/menu.scss'
```
I think it's a good practice to leave the trailing comma

interlineae replied:
This was a JSHint suggestion, but looking into it further, it seems to have been made only out of respect to old versions of IE. Reading up on 'final commas' (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Trailing_commas), trailing commas are indeed good practice!

interlineae replied:
Committed https://github.com/Automattic/WP-Job-Manager/pull/2401/commits/229e6a4f410620d5e9b495cc725f8ca6e7e3a1d0

aaronfc reviewed (commented):
Looks good! Left some comments.

I am adding @renatho so he can give feedback if there is a better way to do something in frontend (assets loading, icon, etc).

renatho commented on assets/css/admin-notices.scss:
```diff
@@ -7,10 +11,22 @@
-	padding: 18px 24px;
+	padding: 10px 12px;
+
+	&.wpjm-admin-notice--error {
```
Just a suggestion to make use of the sass features, keeping it a little cleaner. The same applies to the next classes. ;)

```suggestion
	&--error {
```

interlineae replied:
Committed https://github.com/Automattic/WP-Job-Manager/pull/2401/commits/56386a5274e65a9c5abae0ac8e5e8f7ff62b1401

renatho commented on assets/css/admin-notices.scss:
```diff
@@ -7,10 +11,22 @@
+		border-left-color: $notice-error;
+	}
+
+	&.wpjm-addon-update--warning {
```
Was this accidental to have a different name?

interlineae replied:
Ah, yep, glossed over that. With the change above this has been removed.

interlineae replied:
Commited https://github.com/Automattic/WP-Job-Manager/pull/2401/commits/56386a5274e65a9c5abae0ac8e5e8f7ff62b1401

renatho commented on assets/js/admin/wpjm-notice-dismiss.js:
```diff
@@ -0,0 +1,41 @@
+		} );
+	};
+
+	document.body.addEventListener( 'click', ( event ) => {
```
I imagine you used this approach because the button is being added later for some reason, right?
If yes, I wonder if we could reduce the `body` scope to something more specific. So we prevent running this function on every click on the page. WDYT?

interlineae replied:
I used the approach in Sensei (copied from Sensei), so if we do updates here, we'll want to take a look at Sensei as well. To tighten this up, are we thinking something like this?

```
const wpjmNotice = document.querySelector('.wpjm-admin-notice' );
if ( wpjmNotice ) {
    wpjmNotice.addEventListener( 'click', ( event ) => {
    ...
}
```

interlineae replied:
Committed https://github.com/Automattic/WP-Job-Manager/pull/2401/commits/eb913659a1f603c37f759cab5c791f44ea082b3a

renatho replied:
Hi @anaemnesis! 👋

Yep! Thank you for the change. But in this case, maybe we don't need the conditional anymore since we're only getting clicks under `.wpjm-admin-notice`, right? Not something that would cause a bug though. 🙂

renatho commented on includes/admin/class-wp-job-manager-admin-notices.php:
```diff
@@ -43,14 +46,15 @@ public static function init() {
-	 *
 	 * @param string $notice Name of the notice.
+	 *
+	 * @since 1.32.0
```
Out of curiosity, is there any special reason to move the `@since` to the last item?

interlineae replied:
I haven't been able to find anything suggesting the order is important.

This was done by PHPStorm's linter, and I decided to keep it as it made sense to me in terms of consistency. When I compare, for instance, with Sensei's [class-sensei-admin-notices.php](https://github.com/Automattic/sensei/blob/trunk/includes/admin/class-sensei-admin-notices.php), the `@since` isn't consistently placed in doc blocks (that I can see): sometimes it's last, sometimes it's first.

aaronfc replied:
Yeah, I would disable these "automatic" changes in your PHPStorm configuration. Imagine that VSCode would do the same but putting the since first. We would be "fighting" each other in each change.

aaronfc replied:
I reverted these `@since` changes in my PR [here](https://github.com/Automattic/WP-Job-Manager/pull/2398) (still in draft at the time of writing this).

renatho commented on includes/admin/class-wp-job-manager-admin-notices.php:
```diff
@@ -317,16 +332,40 @@ private static function generate_notice_from_updates( $updates ) {
+			$notice['actions'] = [];
+		}
+
+		$notice_class = '';
```
Maybe it could be an array? So we don't worry about the spacings while when adding other classes. :)
So when using, we just `implode`.

interlineae replied:
Committed https://github.com/Automattic/WP-Job-Manager/pull/2401/commits/4cc876d2ea4b6c27f8079a8886540a29cfd44466

renatho reviewed (commented):
Hey! Adding some comments here too. 😉

interlineae committed:
Addres PHP linter notices

interlineae committed:
Default `$is_dismissible` to true

Defaults `$is_dismissible` to true and removes the 'hard-coded' `is_dismissible` CSS class.

interlineae committed:
Remove skeleton classes

We won't be needing the `get_dismissed_notices` and `save_dismissed_notices` methods after all, so removing.

interlineae committed:
Add final comma back to webpack

Adds the final comma back to the `files` const as this is good practice https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Trailing_commas

interlineae committed:
Update notice Sass classes

It doesn't make sense to have `$.wpjm-admin-notice--*` when we're already styling within `.wpjm-admin-notice`.

interlineae committed:
Initialise `$notice_class` as an empty array rather than string

Initialising as an array means we can use `implode` to output our string of classes, without having to worry about the spacing when adding classes.

interlineae committed:
Make $notice_class['style'] fall back to 'info'

If $notice_class['style'] is empty, we don't want it with grey styling, but with info styling.

interlineae committed:
Ensure $notice['style]' is one of expected type

If $notice['style'] is not error, warning, success, or info, fallback to info.

Also removed unneeded action related to previous skeleton code

interlineae committed:
Add WPJM notice icon

- Adds wpjm-logo.png
- Includes the logo in WPJM icons by default
- Includes additional styling to situate the icon properly

interlineae committed:
Merge remote-tracking branch 'origin/feature/reengaging-customers' into add/admin-notice-modernisation

interlineae committed:
Optimise wpjm-notice-dismiss.js with specificity

wpjm-notice-dismiss.js now targets and looks for event listeners on the `.wpjm-admin-notice` selector specifically (if present), rather than the entire body.

interlineae committed:
Fix harded-coded icon path

Rather than use a hard-codded icon path, use `JOB_MANAGER_PLUGIN_URL` instead. We also won't need `home_url()` anymore.

interlineae committed:
Check that $notice['style'] exists

Before passing $notice['style'] to `in_array`, check that it is set first.

interlineae committed:
Fix PHP linter notices

interlineae committed:
Rename $notice['style'] to $notice['level']

Rename the 'style' attribute to 'level' to match the level of notice, vs the styling of that notice leve.

interlineae committed:
Use higher quality WPJM logo

- Replaces wpjm-logo.png with a higher quality version
- Ensures the logo still only displays at 34x34px

aaronfc reviewed (approved):
Looking good! I tested this in conjunction with my [branch](https://github.com/Automattic/WP-Job-Manager/pull/2398).

I might add there any extra changes if I find anything else that needs to be tweaked.

Good joooob! 🚀
